### PR TITLE
Support running unit tests from another directory

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
When running a test suite, try to change to the directory containing the executable. This allows running a test suite from any directory, and still allow it to access its .datax file as well as data files (generally in tests/data_files) used by individual test cases.

## PR checklist

- [x] **changelog** no (test only)
- [x] **backport** #8421
- [x] **tests** no (test framework feature only)

### CODE REVIEW

File: tests/suites/host_test.function [hunk 1]
```diff
@@ -432,6 +432,38 @@ static void write_outcome_result(FILE *outcome_file,
...
+    defined(_WIN32)
+#define MBEDTLS_HAVE_CHDIR
+#endif
+
+#if defined(MBEDTLS_HAVE_CHDIR)
+/** Try chdir to the directory containing argv0.
+ *
+ * Failures are silent.
+ */
+static void try_chdir(const char *argv0) {
```
Comment (tom-cosgrove-arm) [inline tests/suites/host_test.function:446]: Note: on Windows, compiled by Visual Studio,

```
printf("argv[0] = %s\n", argv0);
```

gives

```
D:\home\tom\projects\mbedtls\master\build>tests\Debug\test_suite_base64.exe
argv[0] = tests\Debug\test_suite_base64.exe
Failed to open test file: .\test_suite_base64.datax
```

or

```
D:\home\tom\projects\mbedtls\master\build>tests\Debug\test_suite_base64
argv[0] = tests\Debug\test_suite_base64
Failed to open test file: .\test_suite_base64.datax
```

i.e. `argv0` has `\`s in it

Comment (gilles-peskine-arm) [reply tests/suites/host_test.function:446]: Good point, I didn't bother with backslash support. I don't want to write complex stuff for Windows because I don't have a system to test it. I'll add a comment to document that it isn't implemented.

Comment (gilles-peskine-arm) [reply tests/suites/host_test.function:446]: Actually, I think I can do something simple that makes sense. Updated. Let me know if it doesn't work.

Comment (gilles-peskine-arm) [reply tests/suites/host_test.function:446]: I'm sure there are plenty of edge cases that don't work, such as file names containing backslashes on a Unix-like environment on Windows, or paths involving drive names. This is a developer convenience, so it doesn't have to handle all cases.

File: tests/suites/host_test.function [hunk 2]
```diff
@@ -432,6 +432,43 @@ static void write_outcome_result(FILE *outcome_file,
...
+    const char *slash = strrchr(argv0, '/');
```
Comment (tom-cosgrove-arm) [inline tests/suites/host_test.function:447]: or you could just do

```
diff --git a/tests/suites/host_test.function b/tests/suites/host_test.function
index 94ae1a07e..c820f805b 100644
--- a/tests/suites/host_test.function
+++ b/tests/suites/host_test.function
@@ -444,7 +444,12 @@ static void write_outcome_result(FILE *outcome_file,
  * Failures are silent.
  */
 static void try_chdir(const char *argv0) {
-    const char *slash = strrchr(argv0, '/');
+    const char *slash = NULL;
+    for (const char *p = argv0; *p != '\0'; p++) {
+        if (*p == '/' || *p == '\\') {
+            slash = p;
+        }
+    }
     if (slash == NULL) {
         return;
     }
```

(yet, technically not the same, but the likelihood of a path on *nix having a `\` in it isn't something that a convenience assist should really worry about)

File: tests/suites/host_test.function (hunk 1 above)
Comment (tom-cosgrove-arm) [reply tests/suites/host_test.function:446]: `try_chdir()` is getting complicated enough with `#if`s that maybe we should switch to

```
diff --git a/tests/suites/host_test.function b/tests/suites/host_test.function
index 06f391fa4..9b53fd142 100644
--- a/tests/suites/host_test.function
+++ b/tests/suites/host_test.function
@@ -432,6 +432,41 @@ static void write_outcome_result(FILE *outcome_file,
     fflush(outcome_file);
 }
 
+#if defined(__unix__) ||                                \
+    (defined(__APPLE__) && defined(__MACH__))
+#define MBEDTLS_CHDIR(dir)      chdir(dir)
+#define PATH_SEPARATOR_CHAR     '/'
+#elif defined(_WIN32)
+#define MBEDTLS_CHDIR(dir)      _chdir(dir)
+#define PATH_SEPARATOR_CHAR     '\\'
+#endif
+
+#if defined(MBEDTLS_CHDIR)
+/** Try chdir to the directory containing argv0.
+ *
+ * Failures are silent.
+ */
+static void try_chdir(const char *argv0) {
+    const char *slash = strrchr(argv0, PATH_SEPARATOR_CHAR);
+    if (slash == NULL) {
+        return;
+    }
+    size_t path_size = slash - argv0 + 1;
+    char *path = mbedtls_calloc(1, path_size);
+    if (path == NULL) {
+        return;
+    }
+    memcpy(path, argv0, path_size - 1);
+    path[path_size - 1] = 0;
+    int ret = MBEDTLS_CHDIR(path);
+    if (ret != 0) {
+        mbedtls_fprintf(stderr, "%s: note: chdir(\"%s\") failed.\n",
+                        __func__, path);
+    }
+    mbedtls_free(path);
+}
+#endif /* MBEDTLS_CHDIR */
+
 /**
  * \brief       Desktop implementation of execute_tests().
  *              Parses command line and executes tests from
```

Comment (gilles-peskine-arm) [reply tests/suites/host_test.function:446]: `chdir`/`_chdir` vs slash-only/slash-or-backslash isn't necessarily the same condition since Unix-like environments on Windows can be confronted with backslashes.

Comment (gilles-peskine-arm) [reply tests/suites/host_test.function]: https://github.com/Mbed-TLS/mbedtls/commit/a5c8361f87921715880bca763cd9fc74f616f8f0 added just enough Windows support to get compilation errors with some mingw builds and runtime error when running unit tests compiled with MSVC and run with cmake's `ctest`. I don't understand why and I don't have the tools to reproduce the problem. So I've completely removed Windows support. If someone wants Windows support, please add it in a follow-up.

File: tests/suites/host_test.function [hunk 3]
```diff
@@ -432,6 +432,43 @@ static void write_outcome_result(FILE *outcome_file,
     fflush(outcome_file);
 }
 
+#if defined(__unix__) ||                                \
+    (defined(__APPLE__) && defined(__MACH__))
+#define MBEDTLS_HAVE_CHDIR
+#endif
+
+#if defined(MBEDTLS_HAVE_CHDIR)
```
Comment (daverodgman) [inline tests/suites/host_test.function:440]: If this is going to be a best-effort function, why not unconditionally define it, and just have it do nothing (apart from maybe print a warning) on platforms like Windows?

Comment (gilles-peskine-arm) [reply tests/suites/host_test.function:440]: What do you mean by “it”? The `try_chdir` function? I guess that could make sense, but then it wouldn't be best-effort anymore, it would be best-effort-if-supported. I guess that's not a big deal either way since I don't expect the function to have more than one call site.

Comment (daverodgman) [reply tests/suites/host_test.function:440]: Yes, `try_chdir`. IMO it's probably slightly simpler to lose the '#if defined ... ' at the call site, but agree not a big deal - not worth iterating for this.

**Review by daverodgman — APPROVED** [review]

**Review by tom-cosgrove-arm — DISMISSED** [review]

**Comment by daverodgman** [comment]
> Backport?

It's not a bugfix, so no IMO.

**Comment by gilles-peskine-arm** [comment]
It's a test framework feature, so yes, I'll make a backport.

File: tests/suites/host_test.function (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/suites/host_test.function:440]: I'm not fully convinced, but I realized this needed rework anyway, so I made this change.

**Comment by daverodgman** [comment]
Note: this doesn't work for use-cases like `valgrind tests/test_suite_xyz` - I guess it is picking up the valgrind executable and using that for the directory to chdir into. I'm not what a good/clean solution to this would be.

**Comment by gilles-peskine-arm** [comment]
That's weird. `valgrind foo/bar` shows `foo/bar` as `argv[0]` to me on Linux.

**Comment by daverodgman** [comment]
> That's weird. `valgrind foo/bar` shows `foo/bar` as `argv[0]` to me on Linux.

It really is wierd:

````
 valgrind $PWD/tests/test_suite_alignment --verbose
<all fine>
valgrind $PWD/tests/test_suite_alignment  --verbose
<all fine>
valgrind --log-file=log $PWD/tests/test_suite_alignment
<all fine>
valgrind --log-file=log $PWD/tests/test_suite_alignment  --verbose
Failed to open test file: ./test_suite_alignment.datax
```

**Comment by gilles-peskine-arm** [comment]
The chdir doesn't happen if there are command line arguments. The intent is that `path/to/test_suite_foo` auto-chdirs, but `path/to/test_suite_foo alternate_file.datax` doesn't. It would be better to only do that if a `.datax` file is specified, and not if there's just command line options, but I didn't think it was worth the additional complexity.

File: tests/suites/main_test.function [hunk 4]
```diff
@@ -237,6 +237,21 @@ int main(int argc, const char *argv[])
...
+     * from another directory (e.g. the project root) and still access
+     * the .datax file as well as data files used by test cases
+     * (typically from tests/data_files).
+     *
+     * Note that we do this before the platform setup (which may access
+     * files such as a random seed). We also do this before accessing
+     * test-specific files such as the outcome file, which is arguably
+     * not desirable and should be fixed later.
+     */
+    if (argc == 1) {
```
Comment (davidhorstmann-arm) [inline tests/suites/main_test.function:251]: Realise that this is best-effort, but this causes execution from a different directory to fail whenever `--verbose` is passed. We would have to check the nature of the argument to test this though, for example check if it starts with `'-'`.

Comment (davidhorstmann-arm) [reply tests/suites/main_test.function:251]: @minosgalanakis just came across this problem

Comment (gilles-peskine-arm) [reply tests/suites/main_test.function:251]: Right, we should do the check on the set of list files rather than the commend line arguments, but IIRC that would have required more refactoring between `main_test.function` and `host_test.function` than I cared to do.
